File: src/irvine/oeis/a248/A248037.java
package irvine.oeis.a248;

import irvine.math.z.Z;
import irvine.oeis.Sequence;
import irvine.util.string.Date;

/**
 * A248037.
 *
 * This program was used in the confirmation of a(15)=3743559068799. It
 * essentially performs an exhaustive search with some additional smarts
 * to avoid evaluating useless candidates.  There are C, Java, and Python
 * versions of this code that are roughly equivalent.
 *
 * We precompute the exact count of triples and halvings for all values less
 * than PRECOMPUTE.  During the actual search we can then truncate the
 * calculation whenever we reach a value less than PRECOMPUTE and simply add
 * the cached counts to the current values.  This saves computing all the
 * way to 1 for every number.
 * 
 * If during the Collatz calculation for n we arrive at some number m,
 * smaller than n and the ratio triples/halves is less than the current
 * best, then the number n can be discarded.
 *
 * For sufficiently large n, n can only be a member of the sequence if
 * n = 7, 15, 27, or 31 (mod 32).  This is because all other values provably
 * have a ratio triples/halvings of less than 0.6.
 *
 * Most values of n less than, say, about 2^50, can be calculated using
 * 63 or 64-bit arithmetic.  We detect when this is not the case by
 * checking if overflow will occur on a tripling step.  When such an
 * overflow is detected, the entire calculation for that number is redone
 * using big number arithmetic.
 *
 * @author Sean A. Irvine
 * @author David Consiglio, Jr.
 */
public class A248037 implements Sequence {

  private static final long MAX_INTERMEDIATE = (Long.MAX_VALUE - 1) / 3;
  private static final int PRECOMPUTE = 1000000;
  private static final Z Z_PRECOMPUTE = Z.valueOf(PRECOMPUTE);
  private final int[] mCacheHalves = new int[PRECOMPUTE];
  private final int[] mCacheTriples = new int[PRECOMPUTE];
  private boolean mVerbose = "true".equals(System.getProperty("oeis.verbose"));
  private long mN = -1;
  private long mLimit = Long.MAX_VALUE - 2;
  private long mBestTriples = -1;
  private long mBestHalves = 1;
  private long mTested = 0;
  private long mExtended = 0;

  private void collatzRatioPrecompute(final int n) {
    // Limit on pre-computation is low enough that this will never overflow
    int halves = 0;
    int triples = 0;
    long m = n;
    while (m > 1) {
      if ((m & 1) == 1) {
        m = 3 * m + 1;
        ++triples;
      }
      do {
        m >>>= 1;
        ++halves;
      } while ((m & 1) == 0);
    }
    mCacheHalves[n] = halves;
    mCacheTriples[n] = triples;
  }

  {
    for (int k = 1; k < PRECOMPUTE; ++k) {
      collatzRatioPrecompute(k);
    }
  }

  private boolean collatzRatio(final long n) {
    long halves = 0;
    long triples = 0;
    long halvesScaled = 0;
    long triplesScaled = 0;
    long m = n;
    while (m >= PRECOMPUTE) {
      if (m < n && triplesScaled < halvesScaled) {
        return false;
      }
      if ((m & 1) == 1) {
        if (m > MAX_INTERMEDIATE) {
          throw new ArithmeticException();
        }
        m = 3 * m + 1;
        ++triples;
        triplesScaled += mBestHalves;
      }
      assert (m & 1) == 0;
      do {
        m >>>= 1;
        ++halves;
        halvesScaled += mBestTriples;
      } while ((m & 1) == 0);
    }
    triples += mCacheTriples[(int) m];
    halves += mCacheHalves[(int) m];
    if (triples * mBestHalves > halves * mBestTriples) {
      mBestTriples = triples;
      mBestHalves = halves;
      return true;
    }
    return false;
  }

  private boolean collatzRatio(final Z n) {
    long halves = 0;
    long triples = 0;
    long triplesScaled = 0;
    long halvesScaled = 0;
    Z m = n;
    while (m.compareTo(Z_PRECOMPUTE) >= 0) {
      if (m.compareTo(n) < 0 && triplesScaled < halvesScaled) {
        return false;
      }
      if (!m.isEven()) {
        m = m.multiply(3).add(1);
        ++triples;
        triplesScaled += mBestHalves;
      }
      assert m.isEven();
      final Z t = m.makeOdd();
      final long aux = m.auxiliary();
      halves += aux;
      halvesScaled += aux * mBestTriples;
      m = t;
    }
    final int u = m.intValueExact();
    triples += mCacheTriples[u];
    halves += mCacheHalves[u];
    if (triples * mBestHalves > halves * mBestTriples) {
      mBestTriples = triples;
      mBestHalves = halves;
      return true;
    }
    return false;
  }

  @Override
  public Z next() {
    if (mN < 1) {
      mN = 1;
      return Z.TWO;
    }
    while (mN <= 8400511) {
      // Early terms do not all obey provable residue classes for later terms
      mN += 2;
      try {
        if (collatzRatio(mN)) {
          if (mVerbose) {
            System.out.println(Date.now() + " Ratio: " + mBestTriples + "/" + mBestHalves + " = " + mBestTriples / (double) mBestHalves);
          }
          return Z.valueOf(mN);
        }
      } catch (final ArithmeticException e) {
        throw new RuntimeException(e); // Should not happen on small inputs
      }
    }
    final long r8 = mN & 7;
    if (r8 == 1 || r8 == 5) {
      mN -= 2;
    }
    while (mN < mLimit) {
      mN += 4;
      final long r = mN & 31;
      if (r == 7 || r == 15 || r == 27 || r == 31) {
        boolean test;
        try {
          test = collatzRatio(mN);
        } catch (final ArithmeticException e) {
          test = collatzRatio(Z.valueOf(mN));
          ++mExtended;
        }
        if (test) {
          if (mVerbose) {
            System.out.println(Date.now() + " Ratio: " + mBestTriples + "/" + mBestHalves + " = " + mBestTriples / (double) mBestHalves);
          }
          return Z.valueOf(mN);
        }
        if (mVerbose && ++mTested == 1000000000L) {
          System.out.println(Date.now() + " Tested: " + mN + " (" + mExtended + " extended)");
          mTested = 0;
          mExtended = 0;
        }
      }
    }
    if (mLimit == Long.MAX_VALUE - 2) {
      throw new UnsupportedOperationException();
    }
    if (mVerbose) {
      System.out.println(Date.now() + " Tested: " + mN + " (" + mExtended + " extended)");
    }
    return null;
  }

  /**
   * Search a specific range.
   * @param args start end triples halves
   */
  public static void main(final String[] args) {
    // Usage: start end besty bestx
    final A248037 seq = new A248037();
    seq.mN = (Long.parseLong(args[0]) | 1) - 2;
    seq.mLimit = Long.parseLong(args[1]);
    seq.mBestTriples = Long.parseLong(args[2]);
    seq.mBestHalves = Long.parseLong(args[3]);
    seq.mVerbose = true;
    Z t;
    while ((t = seq.next()) != null) {
      System.out.println(t);
    }
  }
}

File: c/a248037.c
/**
 * A248037.
 *
 * Compile with:
 *    gcc -Wall -O9 -o a248037 a248037.c -lgmp
 *
 * This program was used in the confirmation of a(15)=3743559068799. It
 * essentially performs an exhaustive search with some additional smarts
 * to avoid evaluating useless candidates.  There are C, Java, and Python
 * versions of this code that are roughly equivalent.
 *
 * We precompute the exact count of triples and halvings for all values less
 * than PRECOMPUTE.  During the actual search we can then truncate the
 * calculation whenever we reach a value less than PRECOMPUTE and simply add
 * the cached counts to the current values.  This saves computing all the
 * way to 1 for every number.
 * 
 * If during the Collatz calculation for n we arrive at some number m,
 * smaller than n and the ratio triples/halves is less than the current
 * best, then the number n can be discarded.
 *
 * For sufficiently large n, n can only be a member of the sequence if
 * n = 7, 15, 27, or 31 (mod 32).  This is because all other values provably
 * have a ratio triples/havings of less than 0.6.
 *
 * Most values of n less than, say, about 2^50, can be calculated using
 * 63 or 64-bit arithmetic.  We detect when this is not the case by
 * checking if overflow will occur on a tripling step.  When such an
 * overflow is detected, the entire calculation for that number is redone
 * using big number arithmetic.
 *
 * @author Sean A. Irvine
 * @author David Consiglio, Jr.
 */
#include <stdio.h>
#include <stdlib.h>
#include <inttypes.h>
#include <limits.h>
#include <gmp.h>
#include <time.h>

// Cache of initial terms (for efficiency only)
const int64_t MAX_INTERMEDIATE = (INT64_MAX - 1) / 3;
const int PRECOMPUTE = 10000000; // < 2^31
mpz_t Z_PRECOMPUTE;
int *cacheHalves;
int *cacheTriples;

// Tracks the best ratio
int64_t bestTriples = -1;
int64_t bestHalves = 1;

// Used for statistics and progress reporting
const char* PROGRESS_FORMAT = "%s Tested: %" PRId64 " (%" PRId64 " extended)\n";
char timeBuffer[27];
int64_t tested = 0;
int64_t extended = 0;

// Workspace for extended precision calculation
mpz_t m;

// Number currently being checked
int64_t candidate = 0;

char *now() {
  time_t timer;
  struct tm* tm_info;
  time(&timer);
  tm_info = localtime(&timer);
  strftime(timeBuffer, 27, "%Y-%m-%d %H:%M:%S ", tm_info);
  return timeBuffer;
}

void collatzRatioPrecompute(int n) {
  // Limit on pre-computation is low enough that this will never overflow
  int halves = 0;
  int triples = 0;
  int64_t m = n;
  while (m > 1) {
    if ((m & 1) == 1) {
      m = 3 * m + 1;
      triples++;
    }
    do {
      m >>= 1;
      halves++;
    } while ((m & 1) == 0);
  }
  cacheHalves[n] = halves;
  cacheTriples[n] = triples;
}

int collatzRatio(int64_t n) {
  int halves = 0;
  int triples = 0;
  int64_t halvesScaled = 0;
  int64_t triplesScaled = 0;
  int64_t m = n;
  while (m >= PRECOMPUTE) {
    if (m < n && triplesScaled < halvesScaled) {
      return 0;
    }
    if ((m & 1) == 1) {
      if (m > MAX_INTERMEDIATE) {
        return -1; // overflow
      }
      m = 3LL * m + 1;
      triples++;
      triplesScaled += bestHalves;
    }
    do {
      m >>= 1;
      halves++;
      halvesScaled += bestTriples;
    } while ((m & 1) == 0);
  }
  // m < PRECOMPUTE < 2^31, hence cast to int is safe
  triples += cacheTriples[(int) m];
  halves += cacheHalves[(int) m];
  // Following multiplies are extremely unlikely to overflow since overflow
  // would require having O(2^32) halvings or triples.  Since current terms
  // are in the hundreds, this will not happen any time soon.
  if (triples * bestHalves > halves * bestTriples) {
    bestTriples = triples;
    bestHalves = halves;
    return 1;
  }
  return 0;
}

int collatzRatioGmp(mpz_t n) {
  int64_t halves = 0;
  int64_t triples = 0;
  int64_t triplesScaled = 0;
  int64_t halvesScaled = 0;
  mpz_set(m, n);
  while (mpz_cmp(m, Z_PRECOMPUTE) >= 0) {
    if (mpz_cmp(m, n) < 0 && triplesScaled < halvesScaled) {
      return 0;
    }
    if (mpz_odd_p(m)) {
      mpz_mul_ui(m, m, 3);
      mpz_add_ui(m, m, 1);
      triples++;
      triplesScaled += bestHalves;
    }
    // We know m must be even at this point, but we want to remove every
    // power of 2.  So count the trailing 0 bits, then shift it down.
    int b = mpz_scan1(m, 1);
    mpz_tdiv_q_2exp(m, m, b);
    halves += b;
    halvesScaled += b * bestTriples;
  }
  // m is less than PRECOMPUTE < 2^31, hence cast to int is safe
  int u = (int) mpz_get_ui(m);
  triples += cacheTriples[u];
  halves += cacheHalves[u];
  if (triples * bestHalves > halves * bestTriples) {
    bestTriples = triples;
    bestHalves = halves;
    return 1;
  }
  return 0;
}

void printBest() {
  double ratio = bestTriples / (double) bestHalves;
  printf("%s Ratio: %" PRId64 "/%" PRId64 " = %.12g\n", now(), bestTriples, bestHalves, ratio);
  fflush(stdout);
}

int64_t next(int64_t end) {
  if (candidate < 1) {
    candidate = 1;
    return 2;
  }
  // Early terms do not all obey provable residue classes for later terms, so
  // (if required) do all these small cases in a separate loop
  int64_t limit1 = end < 8400511 ? end : 8400511;
  while (candidate <= limit1) {
    candidate += 2;
    int res = collatzRatio(candidate);
    if (res > 0) {
      printBest();
      return candidate;
    } else if (res < 0) {
      printf("Something really bad happened, should be impossible\n");
      exit(1);
    }
  }
  // Main loop for searching over candidates, candidate is already odd.
  mpz_t n;
  mpz_init(n);
  int64_t r = candidate & 7;
  if (r == 1 || r == 5) {
    candidate -= 2;
  }
  // Candidate is now 3 or 7 (mod 8), means we can increment by 4 each loop
  while (candidate < end) {
    candidate += 4;
    int64_t r = candidate & 31;
    if (r == 7 || r == 15 || r == 27 || r == 31) {
      int test = collatzRatio(candidate);
      if (test < 0) {
        // Overflow detected in calculation, switch to arbitrary precision
        mpz_set_ui(n, candidate);
        test = collatzRatioGmp(n);
        extended++;
      }
      if (test) {
        printBest();
        mpz_clear(n);
        return candidate;
      }
      if (++tested >= 1000000000L) {
        printf(PROGRESS_FORMAT, now(), candidate, extended);
        fflush(stdout);
        tested = 0;
        extended = 0;
      }
    }
  }
  printf(PROGRESS_FORMAT, now(), candidate, extended);
  return 0;
}

/**
 * Search a specific range.
 */
int main(int argc, char **args) {
  if (sizeof(int64_t) != 8) {
    printf("Expected a machine where int64_t was 64-bits!\n");
    return 1;
  }
  if (argc < 5) {
    printf("Usage: a248037 start end triples havles\n");
    return 1;
  }
  // Prepare a cache of the halvings and triples for all small terms
  mpz_init_set_ui(Z_PRECOMPUTE, PRECOMPUTE);
  cacheHalves = malloc(sizeof(int) * PRECOMPUTE);
  cacheTriples = malloc(sizeof(int) * PRECOMPUTE);
  int k;
  for (k = 1; k < PRECOMPUTE; k++) {
    collatzRatioPrecompute(k);
  }
  // Prepare workspace for extended precision calculations
  mpz_init(m);
  // Perform the search over the requested range
  // Make the start an odd number, because we only need consider odd numbers
  candidate = (atoll(args[1]) | 1) - 2;
  int64_t end = atoll(args[2]);
  bestTriples = atoll(args[3]);
  bestHalves = atoll(args[4]);
  printf("%" PRId64 " %" PRId64 " %" PRId64 "/%" PRId64 "\n", candidate, end, bestTriples, bestHalves);
  int64_t t;
  while ((t = next(end)) > 0) {
    printf("%" PRId64 "\n", t);
  }
  return 0;
}
